File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
#BF-> store all elements in array sort and put it back TC-O(N+N log(N)+N)~ O(N LOGN N) SC- O(2N)~(N)
#Optimal- merge sort inplace TC=O(N) SC O(1)
        if(list1==None):
            return list2
        if(list2==None):
            return list1
        if(list1==None and list2==None):
            return list1
        else:
            dummy= ListNode(-1)
            temp=dummy
            while(list1!=None and list2!=None):
                if(list1.val <=list2.val):
                    temp.next=ListNode(list1.val)
                    list1=list1.next
                else:
                    temp.next=ListNode(list2.val)
                    list2=list2.next
                temp=temp.next

            if(list1!= None):
                temp.next=list1
            else:
                temp.next=list2

            return dummy.next
        
    
        
    

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* head=NULL;
        ListNode* tail=NULL;
        ListNode* t1=list1;
        ListNode* t2=list2;
if(list1==NULL) return list2;
        if(list2==NULL) return list1;
        if(t1->val <=  t2->val){
            head=t1;
            tail=t1;
            t1=t1->next;
        }
        else
        {
            head=t2;
            tail=t2;
            t2=t2->next;  
        }
        while(t1 != NULL && t2!=NULL){
            if(t1->val <=  t2->val){
                tail->next=t1;
                tail=tail->next;
                t1=t1->next;
            }
            else{
                tail->next=t2;
                tail=tail->next;
                t2=t2->next;
            }
        }
        if(t1!=NULL){
            tail->next=t1;
        }
        if(t2 !=NULL){
            tail->next=t2;
        }
return head;
    }
};